django-cloudlaunch/cloudlaunch/backend_plugins/cloudman2_app.py

```diff
@@ -135,70 +128,57 @@ def _run_playbook(self, host, pk, user='ubuntu'):
+            inv = Template(f.read())
         with open(inventory_path, 'w') as f:
-            log.info("Creating inventory file {0}".format(inventory_path))
-            f.writelines(INVENTORY_TEMPLATE.substitute(
-                {'master': host, 'user': user}))
+            log.info("Creating inventory file %s", inventory_path)
+            f.writelines(inv.substitute({'host': host, 'user': user}))
         # Run the playbook
         cmd = "cd {0} && ansible-playbook -i inventory other.yml".format(
             repo_path)
-        log.info("Running Ansible with command {0}".format(cmd))
+        log.info("Running Ansible with command %s", cmd)
         p = subprocess.Popen(cmd, stdout=subprocess.PIPE, shell=True)
-        (out, err) = p.communicate()
+        (out, _) = p.communicate()
         p_status = p.wait()
-        log.info("Playbook stdout: %s\nstatus: %s" % (out, p_status))
-        log.info("Deleting pk file {0} needed by Ansible".format(pkf))
+        log.info("Playbook stdout: %s\nstatus: %s", out, p_status)
         if not settings.DEBUG:
+            log.info("Deleting ansible playbook %s", repo_path)
             shutil.rmtree(repo_path)
         return (p_status, out)
 
-    def launch_app(self, provider, task, name, cloud_config,
-                   app_config, user_data):
-        """
-        Handle the app launch process.
-
-        This will:
-        - Perform necessary checks and env setup, most notably create a new
-          key pair.
-        - Launch an instance and wait until ssh access can be established
-        - Run an Ansible playbook to configure the instance
-        """
-        # Implicitly create a new KP for this instance
-        # Note that this relies on the baseVMApp implementation!
-        kp_name = "CL-" + "".join([c for c in name if c.isalpha() or
-                                  c.isdigit()]).rstrip()
-        app_config['config_cloudlaunch']['keyPair'] = kp_name
-        # Launch an instance and check ssh connectivity
-        result = super(CloudMan2AppPlugin, self).launch_app(
-            provider, task, name, cloud_config, app_config,
-            user_data=None)
-        inst = provider.compute.instances.get(
-            result.get('cloudLaunch', {}).get('instance', {}).get('id'))
-        pk = result.get('cloudLaunch', {}).get('keyPair', {}).get('material')
+    def _configure_host(self, name, task, app_config, provider_config):
+        log.debug("Running CloudMan2AppPlugin _configure_host for %s", name)
+        host = provider_config.get('host_address')
+        user = provider_config.get('ssh_user')
+        ssh_private_key = provider_config.get('ssh_private_key')
         timeout = 0
-        while (not self._check_ssh(inst.public_ips[0], pk=pk) or
+        while (not self._check_ssh(host, pk=ssh_private_key, user=user) or
                timeout > 200):
-            log.info("Waiting for ssh on {0}...".format(inst.name))
+            log.info("Waiting for ssh on %s...", host)
```

Comment:
Using the retry decorator from the retrying package could simplify this? e.g.
https://github.com/gvlproject/cloudbridge/blob/8d705b05d107fd25370de534bbdc1bb64ac42e7b/cloudbridge/cloud/providers/aws/resources.py#L354